Example for route mirroring policy

asaha123 opened the pull request:
Commit Message: Example for route mirroring policy
Additional Description: Add a sandbox for route mirroring policy
Risk Level: Low
Testing: Manual/verify script
Docs Changes:
Release Notes: New sandbox for route mirroring policy
Platform Specific Features:

Fixes: https://github.com/envoyproxy/envoy/issues/23119

phlax commented on examples/route-mirroring/Dockerfile-frontenvoy:
```diff
@@ -0,0 +1,12 @@
+ENV DEBIAN_FRONTEND=noninteractive
+
+RUN apt-get -qq update \
+    && apt-get -qq install --no-install-recommends -y curl \
```
im not seeing any need for curl in the container

phlax commented on docs/root/start/sandboxes/route_mirroring.rst:
```diff
@@ -0,0 +1,126 @@
+    $ docker-compose up
+
+
+Terminal 2 ::
```
im not understanding why this is separated here

asaha123 replied:
Do you mean, why we have this in a separate terminal here? When we run `docker-compose up` in Terminal 1, we are not able to execute any more commands as we are not using the detached mode `-d` for `docker-compose`.

phlax replied:
i think rather follow the same pattern that other sandboxes use of brining up the compose services with `-d` and then there is no need for multiple terminals

asaha123 replied:
@phlax i agree that it may be a good idea in general. However, for this sandbox, viewing the logs from the application is the most straightforward to see that the request was also sent to the mirror cluster. Hence, having the separate terminal open is IMHO better than running `docker logs`. 

If you prefer `docker logs` instead of having the terminal open, happy to change it.

phlax replied:
yes - please make this consistent with how the other sandboxes work

asaha123 replied:
https://github.com/envoyproxy/envoy/pull/23336/commits/3621677290368190a909ca7d960b520088cb155c

phlax commented on docs/root/start/sandboxes/route_mirroring.rst:
```diff
@@ -0,0 +1,126 @@
+
+Incoming requests are received by ``front-envoy`` Envoy proxy.
+
+The ``front-envoy.yaml`` envoy configuration configures two routes::
```
rather use a `literalinclude` to include the config (if you grep in the docs for `literalinclude` you will find examples)

phlax commented on examples/route-mirroring/docker-compose.yaml:
```diff
@@ -0,0 +1,57 @@
+
+  service1:
+    build:
+      context: ../shared/tracing
```
please dont use `shared/tracing` im in the process of removing it altogether - not least because it has multiple procs in the container

rather separate out any backend service to its own container (see eg the zipkin sandbox for an example of that)

phlax commented on examples/route-mirroring/front-envoy.yaml:
```diff
@@ -0,0 +1,196 @@
+              "@type": type.googleapis.com/envoy.extensions.filters.http.router.v3.Router
+
+      transport_socket:
+        name: envoy.transport_sockets.tls
```
unless there is a good reason otherwise, i would stick to using http for this sandbox so that the important parts are clearer

phlax reviewed (commented):
thanks for working on this @asaha123 

first pass review inline ...

asaha123 commented:
> thanks for working on this @asaha123
> 
> first pass review inline ...

Thanks @phlax - pushed a few changes addressing most of the points mentioned.

phlax commented:
@asaha123 copying out the advice from CI:

```console


  Please fix your editor to ensure:

      - no trailing whitespace
      - no preceding mixed tabs/spaces
      - all files end with a newline

```

if you can fix up the formatting and address the remaining issues from previous feedback (remove curl and run composition as a daemon) ill review again

/wait

nguyenptk commented on examples/route-mirroring/verify.sh:
```diff
@@ -0,0 +1,20 @@
+#!/bin/bash -e
+
+export NAME=front-proxy
```
please update the NAME here to `route-mirroring`

nguyenptk commented on examples/route-mirroring/front-envoy.yaml:
```diff
@@ -0,0 +1,105 @@
+                route:
+                  cluster: service2
+                  request_mirror_policies:
+                    cluster_header: "x-mirror-cluster"
```
is the `x-mirror-cluster` specific header? or we can name it with other names?

asaha123 replied:
@nguyenptk the header can be anything - but that's what the request must use to select the mirror cluster

nguyenptk commented:
should we consider adding this case to the logging or document to note for the end users?

> During shadowing, the host/authority header is altered such that -shadow is appended. This is useful for logging. For example, cluster1 becomes cluster1-shadow.

But I think the default log should have the `authority` value, ref: https://www.envoyproxy.io/docs/envoy/latest/configuration/observability/access_log/usage#default-format-string

asaha123 commented:
> should we consider adding this case to the logging or document to note for the end users?
> 
> > During shadowing, the host/authority header is altered such that -shadow is appended. This is useful for logging. For example, cluster1 becomes cluster1-shadow.
> 
> But I think the default log should have the `authority` value, ref: https://www.envoyproxy.io/docs/envoy/latest/configuration/observability/access_log/usage#default-format-string

@nguyenptk not quite sure what you are proposing? Do you suggest we enable logging in envoy and attempt to demonstrate the above behavior?

asaha123 commented:
> @asaha123 copying out the advice from CI:
> 
> ```
>   Please fix your editor to ensure:
> 
>       - no trailing whitespace
>       - no preceding mixed tabs/spaces
>       - all files end with a newline
> ```
> 
> if you can fix up the formatting and address the remaining issues from previous feedback (remove curl and run composition as a daemon) ill review again
> 
> /wait

Thanks @phlax - I am not sure I understand this specific error:

> 2022-10-04T04:07:16.4082857Z ERROR (validation failed): examples/route-mirroring/front-envoy.yaml
2022-10-04T04:07:16.4084525Z Failed to parse static_resources field: Failed to parse listeners field: Failed to parse filter_chains field: Failed to parse filters field: Failed to parse virtual_hosts field: Failed to parse routes field: Failed to parse requ

Any ideas?

asaha123 commented:
@phlax Locally running `./ci/verify-examples.sh route-mirroring` works - i  tried a few times. However, it failed in CI:

> 2022-10-05T03:55:34.8879288Z run_examples@ /home/vsts/work/1/s/ci/verify_examples.sh :39 (./route-mirroring)
2022-10-05T03:55:34.8879887Z   > main@ /home/vsts/work/1/s/ci/verify_examples.sh :44
2022-10-05T03:55:34.9830228Z ##[error]Bash exited with code '1'.

Any suggestions for troubleshooting? Thanks!

phlax commented:
i ran this locally by doing the following...

```console
$ docker-compose up --build -d
$ curl http://localhost:8080/service/1
$ curl http://localhost:8080/service/2
$ docker-compose logs service1-mirror
$ docker-compose logs service2-mirror
```

only the first call to the logs had the expected entries - not sure why but currently it seems that the service2 route is not working as expected - running with `verify.sh` also failed

i guess im wondering if the `service2` route is necessary - whether it is showing anything different from the first route

nguyenptk commented:
@asaha123 

>  Do you suggest we enable logging in envoy and attempt to demonstrate the above behavior?

Yes, I do. I think we need to demonstrate this behavior `-shadow` in access_log in `Step 4: See Logs` of document.

asaha123 commented:
> i ran this locally by doing the following...
> 
> ```
> $ docker-compose up --build -d
> $ curl http://localhost:8080/service/1
> $ curl http://localhost:8080/service/2
> $ docker-compose logs service1-mirror
> $ docker-compose logs service2-mirror
> ```
> 
> only the first call to the logs had the expected entries - not sure why but currently it seems that the service2 route is not working as expected - running with `verify.sh` also failed

You will need to specify the expected cluster header for the call to `/service/2` for Envoy to send the request to service2-mirror and hence for the logs to show.

> running with `verify.sh` also failed

That is strange. It works on my box :(

> 
> i guess im wondering if the `service2` route is necessary - whether it is showing anything different from the first route

Yes, to demonstrate the usage of `cluster_header` to select the mirror cluster, I use `service2` as I believe, it keeps things separate nicely - service1 for static mirror cluster and service 2 for cluster header based mirror cluster.

Hope that makes sense?

asaha123 commented:
> > Any suggestions for troubleshooting? Thanks!
> 
> most likely permissions
> 
> CI tests in an env with `chmod -og` to simulate having a strict `umask` set



> @asaha123
> 
> > Do you suggest we enable logging in envoy and attempt to demonstrate the above behavior?
> 
> Yes, I do. I think we need to demonstrate this behavior `-shadow` in access_log in `Step 4: See Logs` of document.

@nguyenptk I made an update in https://github.com/envoyproxy/envoy/pull/23336/commits/df760cf1705ab1f6279187a4e69bede611f86417 to verify the `Host` header in the application.

I believe that's sufficient?

phlax commented:
still fails locally when run with `./verify.sh` script - and at the same place that its failing in CI - looks like something to do with the port assignments - ill review now

phlax commented on docs/root/start/sandboxes/index.rst:
```diff
@@ -70,6 +70,7 @@ The following sandboxes are available:
     rbac
     redis
     skywalking
+    route_mirroring
```
```suggestion
    route-mirror
```

currently there is a mix of hyphenated and underscored names in both docs and the examples, but i think lets prefer hyphenation

also - i think lets use `route-mirror` (rather than `route-mirroring`) throughout as its shorter/simpler

phlax reviewed (commented):
thanks for iterating this @asaha123 - these sandboxes always take quite a few before they are ready to land

i have reviewed everything except the docs (altho i have read them to get the gist of what is being demonstrated)

asaha123 commented on examples/route-mirroring/verify.sh:
```diff
@@ -0,0 +1,20 @@
+
+
+run_log "Test service: localhost:${PORT_PROXY}/service/1"
+wait_for 20 bash -c "responds_with \"Hello from behind Envoy (service 1)!\" http://localhost:${PORT_PROXY}/service/1"
```
> it should then non-quietly grep for whatever is expected - im not following the grep for the Host header - but lets get it cleaned up and we can think what it should check for

@phlax I added this to demonstrate the `Host` header being modified by Envoy when it sends the request to the mirror cluster (https://github.com/envoyproxy/envoy/pull/23336#issuecomment-1268514584)

asaha123 replied:
@phlax I made some changes in https://github.com/envoyproxy/envoy/pull/23336/commits/ba1d2ed00939584cb47f42f531c684b617154a76 to address the comment regarding the steps matching that in the docs. WDYT?

asaha123 commented:
Green build! :rocket:

phlax commented:
> Green build! 

passes locally also now

/docs

phlax commented on docs/root/start/sandboxes/route-mirror.rst:
```diff
@@ -0,0 +1,146 @@
+
+.. literalinclude:: _include/route-mirror/front-envoy.yaml
+   :language: yaml
+   :lines: 7-37
```
lets reduce the displayed lines - i would probably start on the `virtual_hosts` line - but i think at most 2-3 lines before and 1 after the highlighed text

lets also add a caption with download (you should be able to find existing examples)

phlax commented on docs/root/start/sandboxes/route-mirror.rst:
```diff
@@ -0,0 +1,146 @@
+    $ docker-compose build
+    $ docker-compose up -d
+
+
```
unnecessary whitespace

phlax commented on docs/root/start/sandboxes/route-mirror.rst:
```diff
@@ -0,0 +1,146 @@
+
+.. code-block:: console
+
+
```
unnecessary whitespace

phlax commented on docs/root/start/sandboxes/route-mirror.rst:
```diff
@@ -0,0 +1,146 @@
+    $ pwd
+    envoy/examples/route-mirror
+    $ docker-compose ps
+
```
unnecessary whitespace

phlax commented on docs/root/start/sandboxes/route-mirror.rst:
```diff
@@ -0,0 +1,146 @@
+
+    NAME                                COMMAND                  SERVICE             STATUS              PORTS
+    ---------------------------------------------------------------------------------------------------------------------------
+
```
unnecessary whitespace

phlax commented on docs/root/start/sandboxes/route-mirror.rst:
```diff
@@ -0,0 +1,146 @@
+    ---------------------------------------------------------------------------------------------------------------------------
+
+    route-mirror-front-envoy-1       "/docker-entrypoint.…"   front-envoy         running             0.0.0.0:8001->8001/tcp,
+                                                                                                         :::8001->8001/tcp,
```
i think the ports are now incorrect here - also i think dont wrap it

phlax commented on docs/root/start/sandboxes/route-mirror.rst:
```diff
@@ -0,0 +1,146 @@
+    route-mirror-service2-1          "/usr/local/bin/star…"   service2            running (healthy)
+    route-mirror-service2-mirror-1   "/usr/local/bin/star…"   service2-mirror     running (healthy)
+
+Step 2: Demonstrate static mirror cluster name
```
```suggestion
Step 2: Make a request to the statically mirrored route
```

phlax commented on docs/root/start/sandboxes/route-mirror.rst:
```diff
@@ -0,0 +1,146 @@
+
+Step 2: Demonstrate static mirror cluster name
+**********************************************
+
```
i think this needs some preamble

phlax commented on docs/root/start/sandboxes/route-mirror.rst:
```diff
@@ -0,0 +1,146 @@
+The command above sends a request to the ``envoy-front-proxy`` service which forwards the request to
+``service1`` and also sends the request to the service 1 mirror, ``service1-mirror``.
+
+Step 3: See Logs
```
```suggestion
Step 3: View logs for the statically mirrored request
```

phlax commented on docs/root/start/sandboxes/route-mirror.rst:
```diff
@@ -0,0 +1,146 @@
+``Host`` header was modified by Envoy to have a ``-shadow`` suffix in the
+hostname.
+
+Step 4: Demonstrate mirror cluster via header
```
```suggestion
Step 4: Make a request to the route mirrored by request header
```

phlax commented on docs/root/start/sandboxes/route-mirror.rst:
```diff
@@ -0,0 +1,146 @@
+
+Step 3: See Logs
+****************
+
```
preamble

phlax commented on docs/root/start/sandboxes/route-mirror.rst:
```diff
@@ -0,0 +1,146 @@
+``service2`` and also mirrors the request to the cluster named, ``service2-mirror``.
+
+
+Step 4: See Logs
```
```suggestion
Step 5: View logs for the request mirrored by request header
```

phlax commented on docs/root/start/sandboxes/route-mirror.rst:
```diff
@@ -0,0 +1,146 @@
+
+Step 4: See Logs
+****************
+
```
preamble

phlax commented on docs/root/start/sandboxes/route-mirror.rst:
```diff
@@ -0,0 +1,146 @@
+
+If you do not specify the ``x-mirror-cluster`` in the request to ``service2``,
+``service2-mirror`` will not receive the request.
+
```
could you add a `seealso` section with links to any other reference docs and api please

phlax commented on examples/route-mirror/verify.sh:
```diff
@@ -0,0 +1,24 @@
+. "$(dirname "${BASH_SOURCE[0]}")/../verify-common.sh"
+
+
+run_log "Static mirror cluster: localhost:${PORT_PROXY}/service/1"
```
```suggestion
run_log "Make a request to the statically mirrored route"
```

phlax commented on examples/route-mirror/verify.sh:
```diff
@@ -0,0 +1,24 @@
+
+
+run_log "Static mirror cluster: localhost:${PORT_PROXY}/service/1"
+wait_for 20 bash -c "responds_with \"Hello from behind Envoy (service 1)!\" http://localhost:${PORT_PROXY}/service/1"
```
remove `wait_for`

asaha123 replied:
I seem to need it - the test fails without it. I also see why i think - we need some time for the envoy service to be up and running, I changed it to 10s and that seems to be good enough.

phlax replied:
its not so much about the length of time - if you need it, it suggests something is not coming up correctly - Envoy should be up by the time docker-compose returns

asaha123 replied:
What do you think of making use DELAY as documented [here](https://github.com/envoyproxy/envoy/blob/main/examples/DEVELOPER.md#slow-starting-docker-compositions)?

asaha123 replied:
> Envoy should be up by the time docker-compose returns
It likely is, but there is a race here - as we use `docker-compose up -d`.

phlax replied:
same thing - `DELAY` was added originally and i have been removing it in favour of compositions that come up correctly

im struggling to understand why its failing for you it seems to work for me ok - ill review again later and test then

phlax replied:
> but there is a race here

what race ?

asaha123 replied:
> im struggling to understand why its failing for you it seems to work for me ok - ill review again later and test then

That is strange. i have had success rate of 1/20 times so far without the `wait_for`.

asaha123 replied:
FWIW, `sleep 1` is sufficient for the tests to pass locally..

asaha123 replied:
I have pushed https://github.com/envoyproxy/envoy/pull/23336/commits/014290ffe49ddc26761e3698834af3496b923697 to see what happens in CI without `wait_for`.

phlax commented on examples/route-mirror/verify.sh:
```diff
@@ -0,0 +1,24 @@
+run_log "Static mirror cluster: localhost:${PORT_PROXY}/service/1"
+wait_for 20 bash -c "responds_with \"Hello from behind Envoy (service 1)!\" http://localhost:${PORT_PROXY}/service/1"
+
+docker-compose logs service1 | grep --quiet "Host: localhost:${PORT_PROXY}"
```
```suggestion
run_log "View logs for the request mirrored by request header"
docker-compose logs service1 | grep --quiet "Host: localhost:${PORT_PROXY}"
```


... same with steps below ...

phlax reviewed (commented):
great work @asaha123 - this is taking shape

i have reviewed mostly the structure of the docs <> verify.sh script in this review

overall i think we can make the text/steps a bit clearer - but for now i have more focused on what those steps are 

ive also added a few minor format nits etc

phlax commented:
seems to work in CI - not sure if its because your machine is faster or slower - does this happen with any other sandboxes ?

asaha123 commented:
> seems to work in CI - not sure if its because your machine is faster or slower - does this happen with any other sandboxes ?

Not sure..:( I am just a bit surprised that it is consistently passing in CI and almost consistently failing locally for me.
I did see the same behavior locally on another sandbox i tried.

phlax commented:
>  I did see the same behavior locally on another sandbox i tried.

which sandbox ?

im wondering what is specific to your environment - i looked at the composition and all looks good - i had confidence in this pattern as it has always worked for me locally and noone has reported any errors

that said, the composition doesnt actually wait for envoy to be ready - it just relies on the fact that envoy starts quick enough it always is (/has seemed to be)

there is no way in docker-compose to tell it to wait for a service that is not depended upon by any other service to be healthy - the workaround is to add a stub service that does wait for the last container/s to up and then exits

phlax commented:
possibly related - i saw a CI flake today in the `front-proxy` example that looked as tho the cause was envoy not being ready

not sure if envoy startup has slowed down recently - that particular sandbox needs some work anyhow, so could be unrelated

asaha123 commented:
@phlax I believe, it was the `jaeger-tracing` sandbox.

I would be happy for you to merge this PR as it is, as the `wait_for` issue is not specific to this example as you have mentioned in https://github.com/envoyproxy/envoy/pull/23336#issuecomment-1279143845

phlax commented:
> I would be happy for you to merge this PR as it is,

i think we need a little more work on the docs - ill review now

re the more general sandbox issue im keen to track that down so would be good to follow up - if you are around on the Envoy slack we could probably test some things out quite easily

phlax commented on docs/root/start/sandboxes/route-mirror.rst:
```diff
@@ -0,0 +1,146 @@
+    envoy/examples/route-mirror
+    $ docker-compose build
+    $ docker-compose up -d
+
+.. code-block:: console
+
+    $ pwd
+    envoy/examples/route-mirror
```
i think you can just remove these lines

phlax commented on docs/root/start/sandboxes/route-mirror.rst:
```diff
@@ -0,0 +1,146 @@
+Incoming requests are received by ``envoy-front-proxy`` service. The envoy instance running
+in this container is configured with two routes (:download:`front-envoy.yaml <_include/route-mirror/front-envoy.yaml>`):
+
+.. literalinclude:: _include/route-mirror/front-envoy.yaml
```
im not sure about putting the example code here - i think it distracts a bit from the explanation (and doesnt work so well with formatting in at least my browser)

im wondering whether we should keep it and/or whether we might want to put it in the corresponding steps - ie show the static config in the static config step, etc

asaha123 replied:
I took a stab at inlining it. for example:

![image](https://user-images.githubusercontent.com/76932416/197695720-2178b2a1-f28a-40ac-b425-c1ea62eba78a.png)

phlax commented on docs/root/start/sandboxes/route-mirror.rst:
```diff
@@ -0,0 +1,146 @@
+   :emphasize-lines: 5-17
+   :caption: Envoy configuration with route mirror policy
+
+A request for the path ``/service/1`` is forwarded to the ``service1`` cluster.
+In addition, the request is also forwarded to the ``service1-mirror`` cluster.
```
```suggestion
Requests for the path ``/service/1`` are statically mirrored.

Each request handled by the ``service1`` cluster, and in addition, forwarded to the ``service1-mirror`` cluster.
```

phlax commented on docs/root/start/sandboxes/route-mirror.rst:
```diff
@@ -0,0 +1,146 @@
+A request for the path ``/service/1`` is forwarded to the ``service1`` cluster.
+In addition, the request is also forwarded to the ``service1-mirror`` cluster.
+
+A request for the path ``/service/2`` is forwarded to the ``service2`` cluster.
+If a header ``x-mirror-cluster`` is specified in the request, envoy extracts the
+header value and forwards the request to a cluster with the same name, if found.
+For example, if we send a header with a request as ``x-mirror-cluster: service2-mirror``,
+the request will be forwarded to the ``service2-mirror`` cluster.
```
```suggestion
Requests for the path ``/service/2`` are dynamically mirrored according to the presence and value of the ``x-mirror-cluster`` header.

All reqests for this path are  forwarded to the ``service2`` cluster, and are also mirrored to the cluster named in the header.

For example, if we send a header with a request as ``x-mirror-cluster: service2-mirror``,
the request will be forwarded to the ``service2-mirror`` cluster.
```

Im wondering if we need to say something about this pattern only being useful in a trusted environment - im not sure you would want an end-user determining where a request was mirrored, but i can see how it might be useful for a proxy to add this header for upstream processing, for example.

asaha123 replied:
Good point to mention. I will take a stab at it.

phlax commented on docs/root/start/sandboxes/route-mirror.rst:
```diff
@@ -0,0 +1,146 @@
+
+.. code-block:: console
+
+  $ pwd
+  envoy/examples/route-mirror
```
you can remove these lines

asaha123 replied:
I have removed them from the `curl` commands, but kept them for the `docker-compose` commands, as they have to be run from the specific directory.

phlax replied:
we only need this once at the beginning of the sandbox - we can assume that the user is in the correct dir from there on - its kinda pseudo code anyway

phlax commented on docs/root/start/sandboxes/route-mirror.rst:
```diff
@@ -0,0 +1,146 @@
+  $ curl localhost:10000/service/1
+  Hello from behind Envoy (service 1)!
+
+The command above sends a request to the ``envoy-front-proxy`` service which forwards the request to
```
lets move this above the code block - i think it works better as an intro

phlax commented on docs/root/start/sandboxes/route-mirror.rst:
```diff
@@ -0,0 +1,146 @@
+   Host: localhost-shadow:10000
+   192.168.80.6 - - [06/Oct/2022 03:56:22] "GET /service/1 HTTP/1.1" 200 -
+
+The above logs from the ``service1`` and ``service1-mirror`` services show that
```
same with these paragraphs - i think give the explanation first

phlax commented on docs/root/start/sandboxes/route-mirror.rst:
```diff
@@ -0,0 +1,146 @@
+.. code-block:: console
+
+   $ docker-compose logs service1
+
```
i think remove this whitespace

phlax reviewed (commented):
thanks for iterating @asaha123  - i think we may need to iterate once or twice more

one question i have is about responses - im assuming that responses are just ignored from the mirror and that the returned response is always that of the primary service - wondering if we need to say that somewhere explicitly

the other thing im thinking that would be quite helpful are a couple of additional steps showing a request to the dynamic mirror omitting the mirror header and possibly referring to a non-existent cluster

asaha123 commented:
> re the more general sandbox issue im keen to track that down so would be good to follow up - if you are around on the Envoy slack we could probably test some things out quite easily

sure. what timezone and hours are you online usually? I am not in slack, but I can join the channel during UTC+11 daytime hours

asaha123 commented:
@phlax 

> one question i have is about responses - im assuming that responses are just ignored from the mirror and that the returned response is always that of the primary service - wondering if we need to say that somewhere explicitly

> the other thing im thinking that would be quite helpful are a couple of additional steps showing a request to the dynamic mirror omitting the mirror header and possibly referring to a non-existent cluster

yes, agree. 

Also ready for your review now again!

phlax commented:
> I am not in slack, but I can join the channel during UTC+11 daytime hours

if you join and ping me (`@phlax`) i can respond when online

phlax commented on docs/root/start/sandboxes/route-mirror.rst:
```diff
@@ -0,0 +1,213 @@
+Step 6: Missing or invalid cluster name in request header
+*********************************************************
+
+If you do not specify the ``x-mirror-cluster`` in the request to ``service2``,
+a request to ``service1`` will not automatically be mirrored to ``service2-mirror``.
```
```suggestion
If you do not specify the ``x-mirror-cluster`` in the request to ``service2``,
or specify an unknown cluster, the request will not be mirrored but will be handled in the normal way.
```

phlax commented on docs/root/start/sandboxes/route-mirror.rst:
```diff
@@ -0,0 +1,213 @@
+
+Similarly, if we specify a value for the ``x-mirror-cluster`` header such that there
+is no such cluster defined with that name, ``service2-mirror`` will not receive
+the request:
```
i think collapse these so that you do all of the requests in the first code block and then you can separate out checking the logs to a separate step

phlax reviewed (commented):
docs are looking really good @asaha123 - almost there

once we have got the docs down we will need to revisit the `./verify.sh` script to ensure its testing the correct steps

asaha123 commented on docs/root/start/sandboxes/route-mirror.rst:
```diff
@@ -0,0 +1,213 @@
+   :emphasize-lines: 12-17
+   :caption: Envoy configuration with header based route mirror policy :download:`front-envoy.yaml <_include/route-mirror/front-envoy.yaml>`
+
+The pattern to allow a request header to determine the cluster it is mirrored to
+is most useful in a trusted environment. For example, a higher level Envoy instance (or
+another application acting as a proxy) might automatically add this header to requests for upstream
+processing by another Envoy instance configured with request mirror policies.
```
> thinking more about this - there may other situations where you might allow dynamic mirroring according to end users' requests

yeah, i share your ambiguity here. May be it's better to have this warning, than not. I think.

asaha123 commented on docs/root/start/sandboxes/route-mirror.rst:
```diff
@@ -0,0 +1,213 @@
+
+You can also see that for the request to the ``service2-mirror`` service, the
+``Host`` header was modified by Envoy to have a ``-shadow`` suffix in the
+hostname.
```
That's right. I will move it.

asaha123 commented:
@phlax I have addressed the comments.

> once we have got the docs down we will need to revisit the ./verify.sh script to ensure its testing the correct steps

Step 6 - non-existent or wrong cluster header is not covered by the script as it stands now. 

If you think it's useful to add that, what I can think of is using a specific URL to send to service 2 and check that the request is not received by service2-mirror.

phlax commented on examples/route-mirror/verify.sh:
```diff
@@ -0,0 +1,25 @@
+run_log "View logs for the request mirrored by request header"
+docker-compose logs service1 | grep --quiet "Host: localhost:${PORT_PROXY}"
+docker-compose logs service1-mirror | grep --quiet "Host: localhost-shadow:${PORT_PROXY}"
+
```
```suggestion
docker-compose logs service1-mirror | grep GET
```

phlax commented on examples/route-mirror/verify.sh:
```diff
@@ -0,0 +1,25 @@
+
+run_log "View logs for the request mirrored by request header"
+docker-compose logs service2 | grep --quiet "Host: localhost:${PORT_PROXY}"
+docker-compose logs service2-mirror | grep --quiet "Host: localhost-shadow:${PORT_PROXY}"
```
```suggestion
docker-compose logs service2-mirror | grep --quiet "Host: localhost-shadow:${PORT_PROXY}"
docker-compose logs service2-mirror | grep GET
```

phlax commented on docs/root/start/sandboxes/route-mirror.rst:
```diff
@@ -0,0 +1,213 @@
+   to the client. A response returned by the ``service1-mirror`` or the ``service2-mirror``
+   cluster is not sent back to the client.
+
+   This also means that any problems in request processing in the mirror cluster doesn't affect the
```
i updated my suggestion here to include a point about latency

asaha123 commented:
@phlax no worries at all! I have found your suggestions very useful and helped me improve the sandbox documentation.  

Addressed the last suggestions.

phlax commented:
@asaha123 would you mind addressing https://github.com/envoyproxy/envoy/pull/23336/files#r1007144206

atm, the grammar is slightly incorrect - thanks - then we should be gtm

phlax reviewed (approved):
lgtm, thanks @asaha123 - really appreciated

asaha123 commented:
@phlax thank you, is there anything for me to look at related to the failure/

phlax commented:
no i think it was just an existing flake

/retest

asaha123 commented:
Thanks @phlax for all the help!
